sky/packages/sky/lib/gestures/scale.dart

```diff
@@ -0,0 +1,141 @@
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:sky' as sky;
+
+import 'package:sky/gestures/arena.dart';
+import 'package:sky/gestures/recognizer.dart';
+import 'package:sky/gestures/constants.dart';
+
+enum ScaleState {
+  ready,
+  possible,
+  accepted,
+  started
+}
+
+typedef void GestureScaleStartCallback(sky.Point focalPoint);
+typedef void GestureScaleUpdateCallback(double scale, sky.Point focalPoint);
+typedef void GestureScaleEndCallback();
+
+class ScaleGestureRecognizer extends GestureRecognizer {
+  ScaleGestureRecognizer({ PointerRouter router, this.onStart, this.onUpdate, this.onEnd })
+    : super(router: router);
+
+  GestureScaleStartCallback onStart;
+  GestureScaleUpdateCallback onUpdate;
+  GestureScaleEndCallback onEnd;
+
+  ScaleState _state = ScaleState.ready;
+
+  double _initialSpan;
+  double _currentSpan;
+  Map<int, sky.Point> _pointerLocations;
+
+  double get _scaleFactor => _initialSpan > 0 ? _currentSpan / _initialSpan : 1.0;
+
+  void addPointer(sky.PointerEvent event) {
+    startTrackingPointer(event.pointer);
+    if (_state == ScaleState.ready) {
+      _state = ScaleState.possible;
+      _initialSpan = 0.0;
+      _currentSpan = 0.0;
+      _pointerLocations = new Map<int, sky.Point>();
+    }
+  }
+
+  void handleEvent(sky.PointerEvent event) {
+    assert(_state != ScaleState.ready);
+    bool configChanged = false;
+    switch(event.type) {
+      case 'pointerup':
+        configChanged = true;
+        _pointerLocations.remove(event.pointer);
+        break;
+      case 'pointerdown':
+        configChanged = true;
+        _pointerLocations[event.pointer] = new sky.Point(event.x, event.y);
+        break;
+      case 'pointermove':
+        _pointerLocations[event.pointer] = new sky.Point(event.x, event.y);
+        break;
+    }
+
+    _update(configChanged);
+
+    stopTrackingIfPointerNoLongerDown(event);
+  }
+
+  void _update(bool configChanged) {
+    int count = _pointerLocations.keys.length;
+
+    // Compute the focal point
+    sky.Point focalPoint = sky.Point.origin;
+    for (int pointer in _pointerLocations.keys)
+      focalPoint += _pointerLocations[pointer].toOffset();
+    focalPoint = new sky.Point(focalPoint.x / count, focalPoint.y / count);
+
+    // Span is the average deviation from focal point
+    double totalDeviation = 0.0;
+    for (int pointer in _pointerLocations.keys)
+      totalDeviation += (focalPoint - _pointerLocations[pointer]).distance;
+    _currentSpan = count > 0 ? totalDeviation / count : 0.0;
+
+    if (configChanged) {
+      _initialSpan = _currentSpan;
+      if (_state == ScaleState.started) {
+        if (onEnd != null)
+          onEnd();
+        _state = ScaleState.accepted;
+      }
+    }
+
+    if (_state == ScaleState.ready)
+      _state = ScaleState.possible;
+
+    if (_state == ScaleState.possible &&
+        (_currentSpan - _initialSpan).abs() > kScaleSlop) {
+      resolve(GestureDisposition.accepted);
```

Comment:
This calls acceptGesture which calls this function again.  Is that what you intend?